jni/tests/faiss_memory_test.cpp

```diff
@@ -0,0 +1,298 @@
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+#include "faiss_wrapper.h"
+#include "nmslib_wrapper.h"
+
+#include <vector>
+#include <malloc.h>
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "jni_util.h"
+#include "test_util.h"
+#include "faiss/utils/utils.h"
+
+using ::testing::NiceMock;
+using ::testing::Return;
+#define GTEST_COUT std::cerr << "[          ] [ INFO ]"
+
+TEST(FaissHNSWIndexMemoryTest, BasicAssertions) {
+
+    char dataset[] = "dataset/sift/sift_base.fvecs";
+    float* data_load = NULL;
+    unsigned points_num, dim;
+    test_util::load_data(dataset, data_load, points_num, dim);
+
+    GTEST_COUT << "points_num："<< points_num << " data dimension：" << dim << std::endl;
+    float* dataptr = data_load;
+
+    faiss::idx_t numIds = points_num;
+    std::vector<faiss::idx_t> ids(points_num);
+    std::vector<std::vector<float>> vectors;
+    for (int64_t i = 0; i < numIds; ++i) {
+        ids[i] = i;
+
+        std::vector<float> vect;
+        vect.reserve(dim);
+        for (int j = 0; j < dim; ++j) {
+            vect.push_back(*dataptr);
+	    dataptr++;
+        }
+        vectors.push_back(vect);
+    }
+
+    free(data_load);
+    malloc_trim(0);
+
+    std::string indexPath = "tmp/FaissHNSWIndexMemoryTest.faiss";
+    std::string spaceType = knn_jni::L2;
+    std::string index_description = "HNSW32,Flat";  
+    int thread_num = 7;
+    //int efConstruction = 512;
+    //int efSearch = 512;
+    std::unordered_map<std::string, jobject> parametersMap;
+    parametersMap[knn_jni::SPACE_TYPE] = (jobject)&spaceType;
+    parametersMap[knn_jni::INDEX_DESCRIPTION] = (jobject)&index_description;
+    parametersMap[knn_jni::INDEX_THREAD_QUANTITY] = (jobject)&thread_num;
+    //parametersMap[knn_jni::EF_CONSTRUCTION] = (jobject)&efConstruction;
+    //parametersMap[knn_jni::EF_SEARCH] = (jobject)&efSearch;
+
+    // Set up jni
+    JNIEnv *jniEnv = nullptr;
+    NiceMock<test_util::MockJNIUtil> mockJNIUtil;
+
+    EXPECT_CALL(mockJNIUtil,
+                GetJavaObjectArrayLength(
+                        jniEnv, reinterpret_cast<jobjectArray>(&vectors)))
+            .WillRepeatedly(Return(vectors.size()));
+
+    // Create the index
+    knn_jni::faiss_wrapper::CreateIndex(
+            &mockJNIUtil, jniEnv, reinterpret_cast<jintArray>(&ids),
+            reinterpret_cast<jobjectArray>(&vectors), (jstring)&indexPath,
+            (jobject)&parametersMap);
+
+    // Make sure index can be loaded
+    // std::unique_ptr<faiss::Index> index(test_util::FaissLoadIndex(indexPath));
+
+    // Clean up
+    ids.clear();
+    ids.shrink_to_fit();
+    vectors.clear();
+    vectors.shrink_to_fit();
+
+    malloc_trim(0);
+    size_t mem_usage = faiss::get_mem_usage_kb() / (1 << 10);
```

Comment:
Will this be cross-platform compatible?

Reply:
i try to use faiss util function, i checked the code, it is corss-platform compile check but not useful for windows

```
#ifdef __linux__

size_t get_mem_usage_kb() {
    int pid = getpid();
    char fname[256];
    snprintf(fname, 256, "/proc/%d/status", pid);
    FILE* f = fopen(fname, "r");
    FAISS_THROW_IF_NOT_MSG(f, "cannot open proc status file");
    size_t sz = 0;
    for (;;) {
        char buf[256];
        if (!fgets(buf, 256, f))
            break;
        if (sscanf(buf, "VmRSS: %ld kB", &sz) == 1)
            break;
    }
    fclose(f);
    return sz;
}

#else

size_t get_mem_usage_kb() {
    fprintf(stderr,
            "WARN: get_mem_usage_kb not implemented on current architecture\n");
    return 0;
}
```